nxcomp/Log.cpp

```diff
@@ -0,0 +1,126 @@
+/*                                                                        */
+/* Redistribution and use of the present software is allowed according    */
+/* to terms specified in the file LICENSE.nxcomp which comes in the       */
+/* source distribution.                                                   */
+/*                                                                        */
+/* All rights reserved.                                                   */
+/*                                                                        */
+/* NOTE: This software has received contributions from various other      */
+/* contributors, only the core maintainers and supporters are listed as   */
+/* copyright holders. Please contact us, if you feel you should be listed */
+/* as copyright holder, as well.                                          */
+/*                                                                        */
+/**************************************************************************/
+
+
+#include <stdio.h>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <iomanip>
+
+#include "Log.h"
+
+NXLog nx_log;
+
+
+bool NXLog::will_log() const
+{
+	std::map<std::string, NXLogLevel>::const_iterator item = per_file_levels_.find(current_file());
+
+	if ( item != per_file_levels_.end() )
+	{
+		return current_level() >= item->second;
+	}
+	else
+	{
+		return current_level() >= level();
+	}
+}
+
+
+std::string NXLog::stamp_to_string(const NXLogStamp& stamp) const
+{
+	std::ostringstream oss;
+
+	static const char* level_names[] = {
+		"DEBUG",
+		"INFO ",
+		"WARN ",
+		"ERROR",
+		"FATAL"
+	};
+
+	if ( log_level() )
+		oss << ((stamp.level() >=0 && stamp.level() < NXLOG_LEVEL_COUNT ) ? level_names[stamp.level()] : "???") << " ";
+
+	if ( log_time() )
+	{
+		struct timeval timestamp = stamp.timestamp();
+		struct tm timeinfo;
+
+		localtime_r(&timestamp.tv_sec, &timeinfo);
+
+		if ( log_unix_time() )
+		{
+			oss << timestamp.tv_sec;
+		}
+		else
+		{
+			#if __cplusplus >= 201103L 
+			oss << " " << std::put_time(&timeinfo, "%Y/%m/%d %H:%M:%S");
+			#else
+			oss << timestamp.tv_sec;
+			#endif
+		}
+
+		oss << "." << std::setw(3) << std::setfill('0') << (int)(timestamp.tv_usec / 1000) << " ";
+	}
+
+	if ( log_location() )
+		oss << stamp.file() << "/" << stamp.function() << ":" << stamp.line()  << " ";
+
+	if ( log_thread_id() )
+	{
+		if ( thread_name().empty() )
+			oss << pthread_self() << " ";
+		else
+			oss << "[" << thread_name() << "] ";
+	}
+
+	return oss.str();
+}
+
+NXLog& operator<< (NXLog& out, const NXLogStamp& value)
+{
+	out.current_level( value.level() );
+	out.current_file( value.file() );
+
+	if (  out.will_log() )
+	{
+		if ( out.synchronized() )
+		{
+			out.flush(); // Flush the previous line
+			NXLog::per_thread_data *pdt = out.get_data();
+			(*pdt->buffer) << out.stamp_to_string(value);
+		}
+		else
+		{
+			*(out.stream()) << out.stamp_to_string(value) << " ";
```

Comment:
No, because it will not be needed most of the time. These are the cases for flushing:

* When it's been explicitly asked for, with std::flush or std::endl
* When the text written contains a newline in it
* When the buffer gets too large, as not to waste memory if something produces a huge amount of log output without hitting anything else that triggers a flush.
* When it becomes clear that the previous log line is done. Since writing a NXLogStamp starts a new log entry, we know that the previous one, even if unfinished is done. 

This bit of code is for that last case. Most of the time it's not going to do anything because the log entry before it already ended with a \n or endl and got flushed already. To add newlines here we would have to know if the log entry above forgot to add one, and that just doesn't seem to be worth the trouble.

The intention here is to ensure log entries come out in a timely fashion. If the code forgets a \n somewhere it's just going to look a bit ugly until it's fixed.